Add support for getting a key set for ImmutableOpenMap

arteam opened the pull request:
Currently, we support only getting an iterator which is quite cumbersome to work with. Working with a set view is much better, because it bridges the iterator with `Set`.

#76921 added support for getting a stream on a map's values, we can provide a similar facility for keys and simplify a lot of client code that process the map's keys in an iterative way.

arteam committed:
Add support for getting a Java stream over ImmutableOpenMap keys

Currently, we support only getting an iterator which is quite cumbersome
to work with. The Streams API interacts much better with Java Collections.

arteam committed:
Remove the cursor variable

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into add-support-for-keys-stream-on-immutable-open-map

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into add-support-for-keys-stream-on-immutable-open-map

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into add-support-for-keys-stream-on-immutable-open-map

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into add-support-for-keys-stream-on-immutable-open-map

arteam commented:
@elasticmachine run elasticsearch-ci/eql-correctness

arteam committed:
Add tests for ImmutableOpenMapTests#keysStream

arteam committed:
Add more tests for ImmutableOpenMapTests#keysStream

arteam commented:
@elasticmachine run elasticsearch-ci/bwc

arteam commented:
@elasticmachine run elasticsearch-ci/bwc

arteam commented:
@elasticmachine run elasticsearch-ci/bwc

arteam commented:
@elasticmachine  update branch

arteam commented:
@elasticmachine run elasticsearch-ci/bwc

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into add-support-for-keys-stream-on-immutable-open-map

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into add-support-for-keys-stream-on-immutable-open-map

rjernst commented on server/src/main/java/org/elasticsearch/common/collect/ImmutableOpenMap.java:
```diff
@@ -136,6 +136,23 @@ public void remove() {
+    /**
+     * Returns a direct stream over the keys.
+     */
+    public Stream<KType> keysStream() {
```
Instead of add another non-standard method, can we expose the keyset? This would allow operating on ImmutableOpenMap just like a java Map: get the keyset and call stream() on it.

arteam replied:
I think exposing `keySet` would be difficult without additional copying. Exposing a `Stream` allows the consumer to iterate over the keys without copying them to a set. Also, we already added the `stream` method #76921, so I think it would be consistent to have `keyStream` as alternative to `keyIt`.

rjernst replied:
I don't think exposing a keySet would be much different than the keys() method on the underlying ObjectObjectHashMap. The set doesn't need to be materialized, it is just a wrapper around the keys() object.

henningandersen commented on x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java:
```diff
@@ -185,8 +186,7 @@ public void getSnapshotInfo(GetSnapshotInfoContext context) {
         ImmutableOpenMap<String, IndexMetadata> indicesMap = metadata.indices();
-        ArrayList<String> indices = new ArrayList<>(indicesMap.size());
-        indicesMap.keysIt().forEachRemaining(indices::add);
+        List<String> indices = indicesMap.keysStream().collect(Collectors.toList());
```
Using the proposal that Ryan made, this can turn into `List.copyOf(indicesMap.keySet())` from java 10 and until then `new ArrayList<>(indicesMap.keySet())`. That seems better than stream collecting it.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -1448,7 +1448,7 @@ public Metadata build() {
                     }
                 }
-                indexMetadata.getAliases().keysIt().forEachRemaining(allAliases::add);
+                indexMetadata.getAliases().keysStream().forEach(allAliases::add);
```
I think this has no real benefits and runs at a slightly critical time - so would prefer to just keep this one as is to avoid the wrapper here.

henningandersen commented on server/src/test/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesResponseTests.java:
```diff
@@ -53,12 +52,7 @@ protected GetAliasesResponse mutateInstance(GetAliasesResponse response) {
-            indices.add(keys.next());
-        }
-
+        Set<String> indices = aliases.keysStream().collect(Collectors.toSet());
```
This could turn into `aliases.keySet()` instead.

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodesTests.java:
```diff
@@ -309,30 +310,22 @@ private static DiscoveryNode newNode(int nodeId, Map<String, String> attributes,
-                Set<String> ids = new HashSet<>();
-                nodes.getMasterNodes().keysIt().forEachRemaining(ids::add);
-                return ids;
+                return nodes.getMasterNodes().keysStream().collect(Collectors.toSet());
```
This could become `nodes.getMasterNodes().keySet()`.

henningandersen reviewed (commented):
Thanks Artem, I would like to see this done using `keySet` instead. Thanks for working on enhancing code readability.

arteam committed:
Merge branch 'master' into add-support-for-keys-stream-on-immutable-open-map

arteam committed:
Use a method reference for ImmutableOpenMap::builder

arteam committed:
Fix Gradle integration with Intellij IDEA

In #76897 the `hadoop-common` module was renamed to `hadoop-client-ide`, but
the change wasn't reflected in `elasticsearch.ide.gradle` script.
Because of that an IDE import started failing with the
`Task with path ':plugins:repository-hdfs:hadoop-common:shadowJar" not found` error.

Fix the import by setting the correct module name in the `buildDependencyArtifacts` task.

arteam committed:
Provide keySet instead of keyStream

arteam committed:
Rename keysSet to keySet to be consistent with JUC

arteam committed:
Bring back indexMetadata.getAliases().keysIt().forEachRemaining(allAliases::add)

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into add-support-for-keys-stream-on-immutable-open-map

arteam committed:
Revert "Bring back indexMetadata.getAliases().keysIt().forEachRemaining(allAliases::add)"

This reverts commit 9066ef62dce78e736c146ca320eb2e16074511bc.

arteam committed:
Bring back indexMetadata.getAliases().keysIt().forEachRemaining(allAliases::add)

rjernst commented on server/src/main/java/org/elasticsearch/common/collect/ImmutableOpenMap.java:
```diff
@@ -136,6 +138,24 @@ public void remove() {
+     */
+    public Set<KType> keySet() {
+        final Iterator<KType> keysIterator = keysIt();
+        return new AbstractSet<>() {
```
This doesn't look like it provides a complete set view of the keys, there should also be some other methods overriden for contains/etc?

arteam replied:
`AbstractSet` extends `AbstractCollection` already provides skeleton implementations for `contains`, `remove`, `toArray`, etc based on the `iterator` method.

henningandersen replied:
I think overriding `contains` makes sense since it will be very inefficiently implemented otherwise.

rjernst replied:
I think `add` and 'addAll',  should also be overriden to throw `OperationNotSupportedException`. `remove` and `removeAll` as well, but the behavior can be debated. I think we should make this operate just as `Map.keySet()`, which means the returned set should be "live", so removing a key removes the underlying key from the map. The reason I think we should mimic `Map.keySet()` is so that we can work towards actually making `ImmutableOpenMap implement Map`. This will allow consumers to operate on it as a Map. It is a long term idea, but this would be a start.

henningandersen replied:
I think we should leave those methods throwing an `UnsupportedOperationException`, since the `ImmutableOpenMap` is immutable. We could overload them to refine the message but I find it unnecessary, since it is a bug/coding error anyway.

rjernst replied:
That’s a good point, and why I thought overriding add/addAll was necessary. It hadn’t occurred to me that AbstractMap threw UOE for these already, of course it must.

rjernst commented on server/src/main/java/org/elasticsearch/common/collect/ImmutableOpenMap.java:
```diff
@@ -136,6 +138,24 @@ public void remove() {
+        return new AbstractSet<>() {
+            @Override
+            public Iterator<KType> iterator() {
+                return keysIterator;
```
Shouldn't iterator() create a new iterator, ie call keysIt()?

arteam replied:
`keysIterator` is already holds the new iterator from `keysIt()`. But I can also directly call `keysIt` here

arteam replied:
Done!

rjernst reviewed (changes requested):
Looking better!

arteam committed:
Fix testEmptyKeySetWorks check

It returns ints

arteam committed:
Inline the keysIt call

henningandersen reviewed (commented):
One more important detail otherwise looking good.

arteam committed:
Override contains on AbstractSet

arteam commented:
Thank you both!

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  7.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 77897`
